src/GraphQL/Execution/DocumentExecuter.cs

```diff
@@ -302,11 +314,152 @@ protected virtual Operation GetOperation(string operationName, Document document
+        private async Task ExtractFieldAsync(ExecutionContext context, IObjectGraphType rootType, object source,
+            Field field, FieldType fieldType, ConcurrentDictionary<string, object> data, IEnumerable<string> path)
+        {
+            context.CancellationToken.ThrowIfCancellationRequested();
+
+            var name = field.Alias ?? field.Name;
+
+            if (data.ContainsKey(name))
+            {
+                return;
+            }
+
+            if (!ShouldIncludeNode(context, field.Directives))
+            {
+                return;
+            }
+
+            if (CanResolveFromData(field, fieldType))
+            {
+                var result = ResolveFieldFromData(context, rootType, source, fieldType, field, path);
+
+                data.TryAdd(name, result);
+            }
+            else
+            {
+                var result = await ResolveFieldAsync(context, rootType, source, field, path);
+
+                if (result.Skip)
+                {
+                    return;
+                }
+
+                data.TryAdd(name, result.Value);
+            }
+        }
+
+        private bool CanResolveFromData(Field field, FieldType type)
+        {
+            if (field == null || type == null)
+            {
+                return false;
+            }
+
+            if (type.Arguments != null &&
+                type.Arguments.Any())
+            {
+                return false;
+            }
+
+            if (!(type.ResolvedType is ScalarGraphType))
+            {
+                return false;
+            }
+
+            if (type.ResolvedType is NonNullGraphType)
+            {
+                return false;
+            }
+
+            return true;
+        }
+
+        /// <summary>
+        ///     Resolve simple fields in a performant manor
+        /// </summary>
+        private static object ResolveFieldFromData(ExecutionContext context, IObjectGraphType rootType, object source,
+            FieldType fieldType, Field field, IEnumerable<string> path)
         {
-            return fields.ToDictionaryAsync<KeyValuePair<string, Field>, string, ResolveFieldResult<object>, object>(
-                pair => pair.Key,
-                pair => ResolveFieldAsync(context, rootType, source, pair.Value, path.Concat(new[]{pair.Key})));
+            object result = null;
+
+            try
+            {
+                if (fieldType.Resolver != null)
+                {
+                    var rfc = new ResolveFieldContext(context, field, fieldType, source, rootType, null, path);
+
+                    result = fieldType.Resolver.Resolve(rfc);
+                }
+                else
+                {
+                    result = NameFieldResolver.Resolve(source, field.Name);
+                }
+
+                if (result != null)
+                {
+                    var scalarType = fieldType.ResolvedType as ScalarGraphType;
+
+                    result = scalarType?.Serialize(result);
```

Comment:
I think the `ResolvedType` should always be a `ScalarGraphType` here (due to a previous check), though I think technically the `?` here would potentially change the `result` from non-null to null.  Is that the wanted behavior?  Should this be a hard-cast to `ScalarGraphType` since it should always be one or an error thrown after the `as` check?

Reply:
This is mostly due to me using Resharper which somewhat realizes a null can be returned (although it shouldn't because of the previous check) and playing it safe with a null check vs throwing an exception.